src/drivers/magnetometer/bosch/bmm350/BMM350.cpp

```diff
@@ -0,0 +1,683 @@
+	}
+}
+
+void BMM350::UpdateMagParams()
+{
+	uint8_t odr = GetODR(_param_bmm350_odr.get());
+	uint8_t avg = GetAVG(_param_bmm350_avg.get());
+
+	_mag_odr_mode = odr;
+	_mag_avg_mode = avg;
+	_mag_pad_drive = static_cast<uint8_t>(_param_bmm350_drive.get());
+	PX4_DEBUG("Set params odr = %d, avg = %d, drive = %d", _mag_odr_mode, _mag_avg_mode, _mag_pad_drive);
+}
+
+uint8_t BMM350::GetODR(int value)
+{
+	switch (value) {
+	case 0: return ODR_400HZ;
+
+	case 1: return ODR_200HZ;
+
+	case 2: return ODR_100HZ;
+
+	case 3: return ODR_50HZ;
+
+	case 4: return ODR_25HZ;
+
+	case 5: return ODR_12_5HZ;
+
+	case 6: return ODR_6_25HZ;
+
+	case 7: return ODR_3_125HZ;
+
+	case 8: return ODR_1_5625HZ;
+
+	default: return ODR_200HZ;
+	}
+}
+
+hrt_abstime BMM350::OdrToUs(uint8_t odr) {
+ switch (odr) {
+        case ODR_400HZ:
+            return 2500_us;
+        case ODR_200HZ:
+            return 5000_us;
+        case ODR_100HZ:
+            return 10000_us;
+        case ODR_50HZ:
+            return 20000_us;
+        case ODR_25HZ:
+            return 40000_us;
+        case ODR_12_5HZ:
+            return 80000_us;
+        case ODR_6_25HZ:
+            return 160000_us;
+        case ODR_3_125HZ:
+            return 320000_us;
+        case ODR_1_5625HZ:
+            return 640000_us;
+        default:
+            return 5000_us;
+    }
+}
+
+uint8_t BMM350::GetAVG(int value)
+{
+	switch (value) {
+	case 0: return AVG_NO_AVG;
+
+	case 1: return AVG_2;
+
+	case 2: return AVG_4;
+
+	case 3: return AVG_8;
+
+	default: return AVG_2;
+	}
+}
+
+
+int BMM350::probe()
+{
+	for (int i = 0; i < 3; i++) {
+		const uint8_t CMD = RegisterRead(Register::CMD);
+		const uint8_t CHIP_ID = RegisterRead(Register::CHIP_ID);
+
+		PX4_DEBUG("CMD: 0x%02hhX, CHIP_ID: 0x%02hhX", CMD, CHIP_ID);
+
+		if (CHIP_ID == chip_identification_number) {
+			PX4_DEBUG("Found chip");
+			return PX4_OK;
+
+		} else if (CHIP_ID == 0 && CMD == 0) {
+			PX4_DEBUG("Suspended, but found");
+			return PX4_OK;
+		}
+	}
+
+	return PX4_ERROR;
+}
+
+void BMM350::RunImpl()
+{
+	const hrt_abstime now = hrt_absolute_time();
+	ParametersUpdate();
+
+	switch (_state) {
+	case STATE::RESET: {
+			RegisterWrite(Register::CMD, SOFT_RESET);
+			_reset_timestamp = now;
+			_state = STATE::WAIT_FOR_RESET;
+			perf_count(_reset_perf);
+			ScheduleDelayed(10_ms);
+		}
+		break;
+
+	case STATE::WAIT_FOR_RESET: {
+			uint8_t chipId;
+
+			if ((chipId = RegisterRead(Register::CHIP_ID)) == chip_identification_number) {
+				UpdateMagOffsets();
+				RegisterWrite(Register::OTP_CMD, PWR_OFF_OTP);
+				PX4_DEBUG("After reset chip id = %i", chipId);
+				_state = STATE::AFTER_RESET;
+				ScheduleDelayed(10_ms);
+
+			} else {
+				_state = STATE::RESET;
+				ScheduleDelayed(30_ms);
+			}
+		}
+		break;
+
+	case STATE::AFTER_RESET:
+		uint8_t chipId;
+
+		if (((chipId = RegisterRead(Register::CHIP_ID)) == chip_identification_number)) {
+
+			// Prep self test
+			RegisterWrite(Register::PMU_CMD, PMU_CMD_SUSPEND);
```

Comment:
This looks strange. The chip should always be in SUSPEND after reset, so there is no point in putting it into suspend again. Instead of this, I'd suggest just checking in WAIT_FOR_RESET that the chip actually is in SUSPEND (as this indicates that the reset has succeeded).